test/invariant/handlers/TimeBasedRateHandler.sol

```diff
@@ -1,46 +1,96 @@
 
 import { DSRAuthOracle } from "lib/xchain-dsr-oracle/src/DSRAuthOracle.sol";
 import { IDSROracle }    from "lib/xchain-dsr-oracle/src/interfaces/IDSROracle.sol";
 
-contract TimeBasedRateHandler is StdCheats, StdUtils {
+contract TimeBasedRateHandler is HandlerBase, StdCheats {
 
     uint256 public dsr;
-    uint256 public rho;
 
     uint256 constant ONE_HUNDRED_PCT_APY_DSR = 1.000000021979553151239153027e27;
 
     DSRAuthOracle public dsrOracle;
 
-    uint256 public setRateCount;
+    uint256 public setPotDataCount;
+    uint256 public warpCount;
 
-    constructor(DSRAuthOracle dsrOracle_) {
+    constructor(PSM3 psm_, DSRAuthOracle dsrOracle_) HandlerBase(psm_) {
         dsrOracle = dsrOracle_;
     }
 
     // This acts as a receiver on an L2.
-    function setPotData(uint256 newDsr, uint256 newRho) external {
+    // TODO: Discuss if rho should be set to a value between last rho and block.timestamp.
+    //       This was the original approach but was causing the conversion rate to decrease.
+    function setPotData(uint256 newDsr) external {
+        // 1. Setup and bounds
         dsr = _bound(newDsr, 1e27, ONE_HUNDRED_PCT_APY_DSR);
-        rho = _bound(newRho, rho,  block.timestamp);
+
+        uint256 rho = block.timestamp;
 
         // If chi hasn't been set yet, set to 1e27, else recalculate it in the same way it would
-        // happen during a refresh.
-        uint256 rate = dsrOracle.getConversionRate();
+        // happen during a refresh at `rho`
+        uint256 rate = dsrOracle.getConversionRate(rho);
         uint256 chi  = rate == 0 ? 1e27 : rate;
 
+        // 2. Cache starting state
+        uint256 startingConversion = psm.convertToShares(1e18);
+        uint256 startingValue      = psm.totalAssets();
+
+        // 3. Perform action against protocol
         dsrOracle.setPotData(IDSROracle.PotData({
             dsr: uint96(dsr),
             chi: uint120(chi),
             rho: uint40(rho)
         }));
+
+        // 4. Perform action-specific assertions
+        assertGe(
+            psm.convertToAssetValue(1e18) + 1,
+            startingConversion,
+            "TimeBasedRateHandler/setPotData/conversion-rate-decrease"
+        );
+
+        assertGe(
+            psm.totalAssets() + 1,
+            startingValue,
+            "TimeBasedRateHandler/setPotData/psm-total-value-decrease"
+        );
+
+        // 5. Update metrics tracking state
+        setPotDataCount++;
     }
 
     function warp(uint256 skipTime) external {
-        skip(_bound(skipTime, 0, 45 days));
+        // 1. Setup and bounds
+        uint256 warpTime = _bound(skipTime, 0, 45 days);
+
+        // 2. Cache starting state
+        uint256 startingConversion = psm.convertToShares(1e18);
+        uint256 startingValue      = psm.totalAssets();
+
+        // 3. Perform action against protocol
+        skip(warpTime);
+
+        // 4. Perform action-specific assertions
+        assertGe(
+            psm.convertToAssetValue(1e18),
```

Comment:
Same here.

Reply:
Updated to use `convertToAssetValue` for both